python/lbann/modules.py

```diff
@@ -361,5 +361,148 @@ def forward(self, x, prev_state):
+                              name=self.name+'_hh_matrix'))
+        if len(self.weights) == 3:
+            self.weights.append(
+                lbann.Weights(initializer=lbann.UniformInitializer(min=-1/sqrt(self.size),
+                                                                   max=1/sqrt(self.size)),
+                              name=self.name+'_hh_bias'))
+
+        # Linearity
+        ####Learnable input-hidden weights
+        self.ih_fc = lbann.modules.FullyConnectedModule(3*size, bias=bias,
+                                       weights=self.weights[:2],
+                                       name=self.name + '_ih_fc',
+                                       data_layout=self.data_layout)
+        ###Learnable hidden-hidden weights
+        self.hh_fc = lbann.modules.FullyConnectedModule(3*size, bias=bias,
+                                       weights=self.weights[2:],
+                                       name=self.name + '_hh_fc',
+                                       data_layout=self.data_layout)
+
+    def forward(self, x, prev_state):
+        """Apply GRU step.
+
+        Args:
+            x (Layer): Input.
+            prev_state: State from previous GRU step.
+
+        Returns:
+            (Layer, Layer): The output (out)  and state (hn). 
+                          The state can be passed directly into
+                           the next GRU step.
+
+        """
+        self.step += 1
+        name = '{0}_step{1}'.format(self.name, self.step)
+
+
+        fc1 = self.ih_fc(x)   #input_fc
+        fc2 = self.hh_fc(prev_state)  #hidden_fc
+
+
+        # Get gates and cell update
+        fc1_slice = lbann.Slice(fc1,
+                            slice_points=_str_list([0, self.size, 2*self.size, 3*self.size]),
+                            name=name + '_fc1_slice',
+                            data_layout=self.data_layout)
+        Wir_x = lbann.Identity(fc1_slice, name=name + '_Wrx',
+                           data_layout=self.data_layout)
+        Wiz_x = lbann.Identity(fc1_slice, name=name + '_Wzx',
+                           data_layout=self.data_layout)
+        Win_x = lbann.Identity(fc1_slice, name=name + '_Wnx',
+                           data_layout=self.data_layout)
+
+        fc2_slice = lbann.Slice(fc2,
+                            slice_points=_str_list([0, self.size, 2*self.size, 3*self.size]),
+                            name=name + '_fc2_slice',
+                            data_layout=self.data_layout)
+        Whr_prev = lbann.Identity(fc2_slice, name=name + '_Wrh',
+                           data_layout=self.data_layout)
+        Whz_prev = lbann.Identity(fc2_slice, name=name + '_Wzh',
+                           data_layout=self.data_layout)
+        Whn_prev = lbann.Identity(fc2_slice, name=name + '_Wnh',
+                           data_layout=self.data_layout)
+        
+        rt = lbann.Sigmoid(lbann.Add([Wir_x,Whr_prev]), name=name + '_reset_gate',
```

Comment:
We need to specify the data layout for the add layer, and make similar changes to all the layers below. It would also be nice to make sure they are all named.